### PR TITLE
add recurring algo order handlers

### PR DESCRIPTION
**Pull request by arifjmamun**

### CODE REVIEW

**Comment by dmytroshch** [comment]
Let's use existed `algo...` messages and invoke related recurring ao handlers from ao handlers?

**Comment by dmytroshch** [comment]
https://github.com/bitfinexcom/bfx-hf-server/blob/master/lib/ws_servers/api/handlers/on_algo_order_submit.js#L31-L33
check if `id === 'bfx-hf-recurring'`, if it is, just invoke this [handler](https://github.com/bitfinexcom/bfx-hf-server/pull/217/files#diff-1bf81916292b69f80b8c687a1ad02cef83453f0501cd6e9c6acc07eec809d3bd) and save ao to local db (algos.json)

**Comment by dmytroshch** [comment]
or, for example, for cancelling
https://github.com/bitfinexcom/bfx-hf-server/blob/master/lib/ws_servers/api/handlers/on_algo_order_cancel.js#L25
check if requested order is recurring ao (in local db) and invoke [hanlder](https://github.com/bitfinexcom/bfx-hf-server/pull/217/files#diff-609c7271e20feeecc93674465d59d73b84758a2f3ea719b1498f46988b81bfc9)

**Comment by arifjmamun** [comment]
> also, for example, for cancelling https://github.com/bitfinexcom/bfx-hf-server/blob/master/lib/ws_servers/api/algos/algo_worker.js#L296 check if requested order is recurring ao (in local db) and invoke [handler](https://github.com/bitfinexcom/bfx-hf-server/pull/217/files#diff-609c7271e20feeecc93674465d59d73b84758a2f3ea719b1498f46988b81bfc9)

@tarcisiozf is it make sense to call other handlers conditionally based on `aoID === 'bfx-hf-recurring'`? the recurring AOs will live outside of hf-server/hf-hosted, still do we need to store them locally in `algos.json`? Is it going to be better if it merged into existing algo handlers?

File: lib/ws_servers/api/handlers/on_recurring_algo_order_load_list.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  const { d, restURL } = server
+  const [, authToken, payload] = msg
+  const { apiKey, apiSecret } = ws.getCredentials()
+  const validRequest = validateParams(ws, {
+    authToken: { type: 'string', v: authToken },
+    payload: { type: 'object', v: payload }
+  })
+
+  if (!validRequest) {
+    return
```
Comment (tarcisiozf) [inline lib/ws_servers/api/handlers/on_recurring_algo_order_load_list.js:75]: I think is useful to send an error message

**Comment by dmytroshch** [comment]
please update algo lib to [v4.17.0](https://github.com/bitfinexcom/bfx-hf-algo/releases/tag/v4.17.0) and use `Recurring.id` instead of 'bfx-recurring'

**Comment by arifjmamun** [comment]
> 
please check the file for algo lib v4.17.0: https://github.com/bitfinexcom/bfx-hf-server/pull/217/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519

**Review by tarcisiozf — APPROVED** [review]

**Review by tarcisiozf — CHANGES_REQUESTED** [review]

File: lib/ws_servers/api/index.js [hunk 2]
```diff
@@ -99,6 +105,12 @@ module.exports = class APIWSServer extends WSServer {
         'algo_order_params.get': onAlgoOrderParamsRequest,
         'algo_order_params.remove': onAlgoOrderParamsRemove,
 
+        'recurring_algo_order.submit': onRecurringAlgoOrderSubmit,
```
[the remark this answers is not in the document]
Comment (arifjmamun) [reply lib/ws_servers/api/index.js:108]: yes, we don't need to expose these additional messages. I am going to update it.

File: lib/ws_servers/api/handlers/on_recurring_algo_order_load_list.js [hunk 3]
```diff
@@ -30,11 +30,37 @@ const _getAllRecurringAlgoOrderList = async (rest) => {
     page += 1
   }
 
-  return algoOrders.map(mapRecurringAlgoOrderState)
+  const { AlgoOrder } = algoDB
+
+  return algoOrders.map((ao) => {
+    const mappedRecurringAO = mapRecurringAlgoOrderState(ao)
+
+    AlgoOrder.set({ ...mappedRecurringAO, state: JSON.stringify(mappedRecurringAO.state) })
```
Comment (arifjmamun) [inline lib/ws_servers/api/handlers/on_recurring_algo_order_load_list.js:38]: it is **asynchronous** method and its need to be **awaited**.

**Review by arifjmamun — COMMENTED** [review]
Looks good!

**Comment by dmytroshch** [comment]
Works properly on FE

**Review by tarcisiozf — APPROVED** [review]

**Pull request merged**